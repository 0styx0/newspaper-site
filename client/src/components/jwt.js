import fetchFromApi from '../helpers/fetchFromApi';

let jwt = {};

async function setJWT() {

    return await fetchFromApi('userStatus')
        .then(data => data)
        .then(data => data.json())
        .then((json) => {
            jwt = json;
            return jwt;
        });
}

export {jwt, setJWT};