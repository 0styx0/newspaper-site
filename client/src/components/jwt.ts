
interface Jwt {
    id: string;
    level: number;
    email: string;
}

let jwt: Jwt = {
    id: '',
    level: 1,
    email: ''
};

function getJWT(): Jwt {

    return JSON.parse(window.localStorage.getItem('jwt') || JSON.stringify(jwt));
}

async function setJWT(newJwt: Jwt) {

    window.localStorage.setItem('jwt', JSON.stringify(newJwt));
    return newJwt;
}

export {getJWT, setJWT, Jwt};