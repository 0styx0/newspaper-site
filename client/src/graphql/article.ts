import gql from 'graphql-tag';

const ArticleQuery = gql`
    query ArticleQuery($issue: Int!, $url: String!) {
        articles(issue: $issue, url: $url) {
            id,
            article,
            canEdit,
            tags,
            comments {
                id
                content,
                dateCreated,
                canDelete,
                author {
                    fullName,
                    profileLink
                }
            }
        }
    }

`;

const ArticleCreate = gql`
    mutation ArticleCreate($tags: [String], $url: String!, $article: String!) {
        createArticle(tags: $tags, url: $url, article: $article) {
            issue
            url
        }
    }
`;

export {
    ArticleQuery,
    ArticleCreate
};
