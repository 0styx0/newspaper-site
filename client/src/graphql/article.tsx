import { gql } from 'react-apollo';

const ArticleQuery = gql`
    query ArticleQuery($issue: Int!, $url: String!) {
        articles(issue: $issue, url: $url) {
            id,
            article,
            canEdit,
            comments {
                id
                content,
                dateCreated,
                canEdit,
                author {
                    fullName,
                    profileLink
                }
            }
            tags {
                all
            }
        }
    }

`;

export {
    ArticleQuery
};
