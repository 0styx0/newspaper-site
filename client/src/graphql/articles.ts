import gql from 'graphql-tag';

const ArticleQuery = gql`

    query ArticleQuery($issue: ID) {
        issues(limit: 1, num: $issue) {
          max
          num
          articles {
            tags
            url
            id
            displayOrder
            dateCreated
            views
            author {
                fullName
                profileLink
            }
          }
        }
    }
`;

const ArticlePreviewIssueQuery = gql`

    query ArticleIssueQuery($issue: ID) {
        issues(limit: 1, num: $issue) {
          articles {
            url
            issue
            displayOrder
            views
            lede
            images(slide: true) {
                url
            }
          }
        }
    }
`;

const ArticlePreviewTagQuery = gql`

    query ArticleTagQuery($issue: Int, $tag: String) {
        articles(issue: $issue, tag: $tag) {
          url
          issue
          displayOrder
          views
          lede
          images(slide: true) {
              url
          }
        }
    }
`;

const ArticleUpdate = gql`
    mutation ArticleUpdate($data: [UpdateArticle], $password: String) {
        updateArticles(data: $data, password: $password) {
            id
            tags
            displayOrder
        }
    }
`;

const EditArticle = gql`
    mutation EditArticle($id: ID, $article: String) {
        editArticle(id: $id, article: $article) {
            id
            tags
            displayOrder
        }
    }
`;

const ArticleDelete = gql`

    mutation deleteArticles($ids: [ID], $password: String!) {
        deleteArticles(ids: $ids, password: $password) {
            id
        }
    }

`;

export {
    ArticleQuery,
    ArticlePreviewIssueQuery,
    ArticlePreviewTagQuery,
    ArticleUpdate,
    EditArticle,
    ArticleDelete
};