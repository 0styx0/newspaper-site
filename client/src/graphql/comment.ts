import { gql } from 'react-apollo';

const CommentCreate = gql`

    mutation CommentCreate($artId: ID!, $content: String!) {
        createComment(artId: $artId, content: $content) {
            id
        }
    }
`;

export {
    CommentCreate
};