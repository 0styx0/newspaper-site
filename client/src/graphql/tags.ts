import gql from 'graphql-tag';

const TagQuery = gql`

    query allTags {
        allTags
    }
`;

const TagCreate = gql`

    mutation createTag($tag: String!) {
        createTag(tag: $tag) {
            tag
        }
    }
`;

export {
    TagQuery,
    TagCreate
};
