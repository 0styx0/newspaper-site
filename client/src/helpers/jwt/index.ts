
interface Jwt {
    id: string;
    level: number;
    profileLink: string;
}

let jwt: Jwt = {
    id: '',
    level: 0,
    profileLink: ''
};

/**
 * @return parsed jwt's body, if the jwt exists in window.localStorage,
 *  else returns {jwt} (basically a placeholder)
 */
function getJWT(): Jwt {

    const encodedToken = window.localStorage.getItem('jwt');

    if (encodedToken) {

        const encodedBody = encodedToken.split('.')[1];
        const jsonString = atob(encodedBody);
        return JSON.parse(jsonString);

    } else {
        return jwt;
    }
}

/**
 * Stores the jwt
 *
 * @param newJwt - properly encoded json web token
 */
function setJWT(newJwt: string) {

    window.localStorage.setItem('jwt', newJwt);
    return newJwt;
}

export {getJWT, setJWT, Jwt};