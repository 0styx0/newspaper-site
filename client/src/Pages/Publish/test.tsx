import * as React from 'react';
import '../../tests/setup.mocks';
import PublishContainerWithApollo, { PublishContainer, Props, State } from './container';
import * as renderer from 'react-test-renderer';
import * as sinon from 'sinon';
import casual from '../../tests/casual.data';
import { submitForm, setupComponent, setInput, setSelectByName } from '../../tests/enzyme.helpers'
import { ReactWrapper } from 'enzyme';
import setFakeJwt from '../../tests/jwt.helper';
import { getJWT } from '../../helpers/jwt/index';
import { ApolloProvider } from 'react-apollo';
const wait = require('waait');

import mockGraphql, { createQuery, createMutation, mountWithGraphql } from '../../tests/graphql.helper';
import { ArticleCreate } from '../../graphql/article';
import { TagCreate, TagQuery } from '../../graphql/tags';
import { tagQueryMock } from '../../tests/graphql.mocks';
import { renderWithGraphql } from '../../tests/snapshot.helper';

setFakeJwt({ level: 3 });


describe('<PublishContainer>', () => {

    function mockCreateArticle() {

        return createQuery(
            ArticleCreate,
            {
                createArticle: {
                    url: casual.articleUrl,
                    issue: casual.issue
                }
            }
        );
    }

    async function setup(graphql = []
    ) {

        return mountWithGraphql(
            graphql,
            <PublishContainerWithApollo history={[]} />
        );
    }

    /**
     * Gives PublishContainer a fake version of state.editor (that has all functionality needed)
     */
    function setFakeEditor(component: PublishContainer, initialContent: string = '') {

        component.setState({

            editor: {
                content: initialContent,
                setContent(content: string) {
                    this.content = content;
                },
                getContent() {
                    return this.content!;
                }
            }
        });
    }

    interface FormData {
        url: string;
        tags: Set<string>;
        article: string;
    }

    function generateFormData(): FormData {

        return {
            url: casual.articleUrl,
            article: casual.article,
            tags: casual.tags
        };
    }

    /**
     * Fills out form with random data
     */
    function fillOutForm(wrapper: ReactWrapper<Props, State>, formData: FormData) {

        const { url, tags, article } = formData;

        const component = setupComponent(wrapper, PublishContainer);
        setFakeEditor(component, article);

        setInput(wrapper, url, 'name');

        setSelectByName(wrapper, 'tags', tags);

        return {
            url,
            article,
            tags: [...tags]
        };
    }
    describe('snapshots', () => {

        async function snap(level: number) {

            const oldLevel = getJWT().level;
            setFakeJwt({ level });

            await renderWithGraphql(
                mockGraphql(
                    [tagQueryMock],
                    <PublishContainerWithApollo />
                )
            )

            setFakeJwt({ level: oldLevel });
        }

        it('renders correctly when can add tag', async () => {
            await snap(3);
        });

        it('does not render option to add tag if not level 2', async () => {
            await snap(1);
        });
    });

    describe('#autoFormat', () => {

        /**
         * Sets up and tests #autoFormat
         *
         * @param content - original, badly formatted content
         * @param expected - properly formatted version of content
         */
        async function testAutoFormat(content: string, expected: string) {

            const wrapper = await setup();
            const component = setupComponent(wrapper, PublishContainer);
            setFakeEditor(component, content);
            component.autoFormat();
            const newEditorContents = component.state.editor!.getContent();

            expect(newEditorContents).toBe(expected);
        }

        it(`adds an <h0> if none exists and there's text`, async () => {

            const author = casual.full_name;
            const title = casual.title;

            await testAutoFormat(`<h4>${title}</h4><p>${author}</p>`, `<h1>${title}</h1><h4>${author}</h4>`);
        });

        it('adds an <h4> if none exists', () => {

            const author = casual.full_name;
            const title = casual.title;

            testAutoFormat(`<h1>${title}</h1><p>${author}</p>`, `<h1>${title}</h1><h4>${author}</h4>`);
        });

        it('adds an <h1> and <h4> if none exists', () => {

            const author = casual.full_name;
            const title = casual.title;

            testAutoFormat(`<p>${title}</p><strong>${author}</strong>`, `<h1>${title}</h1><h4>${author}</h4>`);
        });

        it(`doesn't mess with any html other than first 2 tags`, () => {

            const author = casual.full_name;
            const title = casual.title;

            testAutoFormat(
                `<p>${title}</p><strong>${author}</strong><em>This should not change</em>`,
                `<h1>${title}</h1><h4>${author}</h4><em>This should not change</em>`
            );
        });
    });

    describe('#onTagChange', () => {

        it(`toggles addTag input when 'other' is selected`, async () => {

            const wrapper = await setup();
            setupComponent(wrapper, PublishContainer);

            expect(wrapper.find('input[name="addTag"]').length).toBe(0);

            setSelectByName(wrapper, 'tags', ['other']);

            expect(wrapper.find('input[name="addTag"]').length).toBe(1);
        });

        it('submits new tag to createTag', async () => {

            let newTag = casual.word;

            const formData = generateFormData();
            formData.tags = [newTag];

            const wrapper = await setup([
                createMutation(
                    ArticleCreate,
                    formData,
                    { createArticle: { issue: casual.randomPositive, url: casual.string}}
                ),
                createMutation(
                    TagCreate,
                    { tag: newTag },
                    { createTag: { tag: newTag }}
                )
                ]);

            const component = setupComponent(wrapper, PublishContainer);
            formData.tags = ['other'];
            fillOutForm(wrapper, formData);

            setInput(wrapper, newTag, 'addTag');

            await submitForm(wrapper);
        });
    });

    describe('#onSubmit', () => {

        it('calls props.createArticle when submit button is clicked', async () => {

            const formData = generateFormData();
            const wrapper = await setup([
                createMutation(
                    ArticleCreate,
                    formData,
                    { createArticle: { issue: casual.randomPositive, url: casual.string}}
                ),
                createQuery(TagQuery, { allTags: Array.from(formData.tags)})
            ]);

            fillOutForm(wrapper, formData);

            await submitForm(wrapper);
        });
    });
});
