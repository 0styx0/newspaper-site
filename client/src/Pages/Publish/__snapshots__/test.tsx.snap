// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PublishContainer> snapshots renders correctly 1`] = `
<section
  className="container "
>
  <h1>
    Publish Story
  </h1>
  <form
    onSubmit={[Function]}
  >
    <label>
      Tags: 
      <select
        multiple={true}
        name="tags"
        required={true}
      >
        <option
          value="opinion"
        >
          Opinion
        </option>
        <option
          value="reaction"
        >
          Reaction
        </option>
        <option
          value="news"
        >
          News
        </option>
        <option
          value="other"
        >
          Other
        </option>
      </select>
      
    </label>
    <label
      onInput={[Function]}
    >
      Page Name: 
      <input
        autoFocus={true}
        maxLength={75}
        name="name"
        pattern="^[\\\\sa-zA-Z0-9_-]+$"
        placeholder="mystory"
        required={true}
        title="Must only contain letters, numbers, -, _, and spaces"
        type="text"
      />
      <span
        className="danger"
      >
        *
      </span>
      <span>
        <abbr
          onClick={[Function]}
          title="This article will be located at tabceots.com/issue/n/story/name_you_enter
                                    (where n is the issue number). Can be up to 75 characters long
                                    (some non-word characters such as spaces may count as more)."
        >
          ?
        </abbr>
        <span />
      </span>
    </label>
    <button
      onClick={[Function]}
      type="button"
    >
      Auto Format
    </button>
    <textarea
      className="changed"
      id="editor"
      name="txtArea"
    />
    <input
      className="submit"
      name="create"
      type="submit"
      value="Submit"
    />
  </form>
</section>
`;
